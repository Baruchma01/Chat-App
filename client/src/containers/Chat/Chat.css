.ChatContainer {
    display: flex;
    flex-wrap: wrap;
    flex-direction: column;
    background: #fff;
    width: 780px;
    margin: 30px auto;
    border-radius: 6px;
    box-shadow: 0 0 6px rgba(0, 0, 0, 0.3);
    overflow: hidden;
    min-height: 530px;
}

.ChatHeader {
    display: flex;
    flex-wrap: wrap;
    justify-content: space-between;
    width: 100%;
    align-items: center;
    flex: 1;
    border-bottom: 1px solid #cfdae1;
}

.Dots {
    display: flex;
    gap: 5px;
    flex-wrap: wrap;
}

.ChatSection {
    display: flex;
    flex-wrap: wrap;
    width: 100%;
    background-color: purple;
    flex: 10;
}

.ConversationList {
    display: flex;
    flex-direction: column;
    flex-wrap: wrap;
    width: 25%;
    background-color: #505d71;
    justify-content: space-between;
}

.ChatArea {
    display: flex;
    flex-direction: column;
    flex-wrap: wrap;
    width: 75%;
    background-color: #fff;
}

ul {
    list-style: none;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    padding: 0;
    margin: 0;
    padding-left: 1rem;
}

.List {
    width: 100%;
    display: flex;
    border-bottom: 2px solid #586476;
    align-items: center;
    gap: 10px;
}

li {
    padding: 0.7rem 0rem;
    color: #fff;
}

.List:hover {
    background-color: #445166;
}

p {
    padding: 0.7rem 0rem;
    color: #fff;
    margin: 0;
}

.ButtomList {
    width: 100%;
    display: flex;
    border-top: 2px solid #586476;
    align-items: center;
    gap: 10px;
    padding-left: 1rem;
}


.Title {
    padding: 10px;
    overflow: hidden;
    line-height: 15px;
    border-bottom: 1px solid #cfdae1;
    border-right: 1px solid #cfdae1;
}

.MessageArea {
    list-style: none;
    display: flex;
    justify-content: start;
    border-bottom: 1px solid #cfdae1;
    border-right: 1px solid #cfdae1;
}

.MsgContainer {
    height: 408px;
}

.InputArea {
    display: flex;
    flex-wrap: wrap;
}

.InputWrapper {
    display: flex;
    flex-wrap: wrap;
    width: 80%;
}

.InputWrapper input {
    width: 100%;
}

.BtnSubmit {
    width: 20%;
    background: #a0b4c0;
    border: none;
    letter-spacing: 1px;
    font-weight: bold;
    color: #fff;
    border-radius: 4px;
    float: right;
    height: 36px;
    line-height: 30px;
    cursor: pointer;
    font-family: 'Open Sans', Arial, sans-serif;
    outline: none;
}

.MessageLi {
    display: flex;
    width: 100%;
    color: #000;
    justify-content: center;
    /* align-items: center; */
    padding: 0;
}

.Msg {
    width: 80%;
    padding-left: 1rem;
    font-size: 14px;
}

.MsgName {
    width: 20%;
    border-right: 1px solid #cfdae1;
    display: flex;
    align-items: center;
    font-size: 14px;
}

.Msg p {
    color: #000;
}