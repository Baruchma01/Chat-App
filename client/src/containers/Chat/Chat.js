import React, { useState, useEffect, useRef } from "react";
import classes from "./Chat.css";
import openSocket from "socket.io-client";
import Spinner from "../../components/Spinner/Spinner";

const socket = openSocket(process.env.REACT_APP_API_URI);
const Chat = ({ location }) => {
  const divRref = useRef(null);
  const [msg, setMsg] = useState("");
  const [roomMessage, setRoomMessage] = useState([]);
  const [activeUsers, setActiveUsers] = useState([]);
  const [currentRoom, setCurrentRoom] = useState("");

  const { username, room } = location.state.user;

  useEffect(() => {
    socket.emit("joinRoom", { username, room });
  }, []);

  useEffect(() => {
    // Get message from the server
    socket.on("message", (message) => {
      setRoomMessage((state) => [...state, message]);
      if (roomMessage.length > 0) {
        divRref.current.scrollIntoView({ behavior: "smooth" });
      }
    });

    // Get room and users
    socket.on("roomUsers", ({ room, users }) => {
      setActiveUsers((state) => users);
      setCurrentRoom((state) => room);
    });
  }, []);

  const submitMessageHandler = (e) => {
    e.preventDefault();
    if (msg.length === 0) {
      alert("error: empty msg");
    } else {
      socket.emit("chatMessage", msg);
      setMsg("");
    }
  };

  const handleMsgChange = (e) => {
    setMsg(e.target.value);
  };

  const greenDot = {
    color: "#82cf85",
  };

  const specialColor = {
    color: "#79889d",
  };

  const liGray = {
    background: "#e4eaee",
  };

  return (
    <>
      {!socket.connected ? (
        <Spinner />
      ) : (
        <div className={classes.ChatContainer}>
          <div className={classes.ChatHeader}>
            <div className={classes.Dots}>
              <i
                className="fa fa-circle fa-xs"
                style={{ color: "#f57e7d" }}
              ></i>
              <i
                className="fa fa-circle fa-xs"
                style={{ color: "#ffc881" }}
              ></i>
              <i
                className="fa fa-circle fa-xs"
                style={{ color: "#82cf85" }}
              ></i>
            </div>
            <div>
              <p style={{ margin: 0 }}>Code Chat</p>
            </div>
            <div>
              <i className="fa fa-expand"></i>
            </div>
          </div>
          <div className={classes.ChatSection}>
            <div className={classes.ConversationList}>
              <ul>
                <div className={classes.List}>
                  <i style={specialColor} className="fa fa-list-alt fa-xs"></i>
                  <li>{currentRoom}</li>
                </div>
                <div className={classes.List}>
                  <i style={specialColor} className="fa fa-user fa-xs"></i>
                  <li>Team chat</li>
                </div>
                <div style={{ width: "100%" }}>
                  {(activeUsers || []).map((user) => (
                    <div className={classes.List} key={user.id}>
                      <i
                        style={greenDot}
                        className="fa fa-circle-o online fa-xs"
                      ></i>
                      <li>{user.username}</li>
                    </div>
                  ))}
                </div>
              </ul>
              <div className={classes.ButtomList}>
                <a href="/" className={classes.Btn}>
                  <button>Leave</button>
                </a>
              </div>
            </div>

            <div className={classes.ChatArea}>
              <div className={classes.Title}>
                <span>Conversation title</span>
              </div>

              <div className={classes.MsgContainer}>
                <div className={classes.singleMsg}>
                  {(roomMessage || []).map((message, index) => {
                    return (
                      <ul
                        key={index}
                        style={liGray}
                        className={classes.MessageArea}
                      >
                        <li className={classes.MessageLi}>
                          <div className={classes.MsgName}>
                            <span className="">{message.username}</span>
                          </div>
                          <div className={classes.Msg}>
                            <p>{message.text}</p>
                            <span className="msg-time">{message.time}</span>
                          </div>
                        </li>
                      </ul>
                    );
                  })}
                </div>
              </div>

              <form className={classes.InputArea}>
                <div className={classes.InputWrapper}>
                  <input
                    type="text"
                    placeholder="Enter Message"
                    value={msg}
                    onChange={handleMsgChange}
                  />
                </div>

                <button
                  className={classes.BtnSubmit}
                  onClick={submitMessageHandler}
                >
                  {" "}
                  Submit
                </button>
              </form>
            </div>
          </div>
        </div>
      )}
    </>
  );
};

export default Chat;
