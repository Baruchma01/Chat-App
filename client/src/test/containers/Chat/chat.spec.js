import * as React from "react";
import { shallow } from "enzyme";
import Chat from "../../../containers/Chat/Chat";
import io from "socket.io-client";


jest.mock("socket.io-client", () => {
  const emit = jest.fn();
  const on = jest.fn();
  const socket = { emit, on };
  return jest.fn(() => socket);
});

describe("<App />", () => {

  beforeEach(() => { 
    io.mockClear();
    io().on.mockClear();
    io().emit.mockClear();
  });

  it("should render Join component with location props", () => {
    const location = {
      state : {
        user: {
          username: 'Baruch',
          room: 'JavaScript'
        }
      }
    }

    const wrapper = shallow(<Chat location={location}/>);
    expect(wrapper).toMatchSnapshot();
  });
});
