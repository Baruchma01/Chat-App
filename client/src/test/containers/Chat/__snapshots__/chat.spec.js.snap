// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render Join component with location props 1`] = `
<Fragment>
  <_default />
  <ToastContainer
    autoClose={false}
    closeButton={[Function]}
    closeOnClick={true}
    draggable={true}
    draggablePercent={80}
    hideProgressBar={false}
    newestOnTop={false}
    pauseOnFocusLoss={true}
    pauseOnHover={true}
    position="top-center"
    role="alert"
    rtl={false}
    transition={[Function]}
  />
  <div
    className="ChangeNickName"
  >
    <a
      className="BtnNickName"
      href="/"
    >
      <button
        className="LogOutBtn"
      >
        Chane Nick Name
      </button>
    </a>
  </div>
</Fragment>
`;
